bigtable_rs/src/bigtable.rs

```diff
@@ -316,42 +315,58 @@ impl BigTableConnection {
 
                 let auth_manager = Some(Arc::new(authentication_manager));
                 Ok(Self {
-                    client: create_client(endpoints, auth_manager, is_read_only),
+                    client: create_client(endpoints, None, auth_manager, is_read_only),
                     table_prefix: Arc::new(table_prefix),
                     timeout: Arc::new(timeout),
                 })
             }
         }
     }
 
-    fn new_with_emulator(
+    pub fn new_with_emulator(
         emulator_endpoint: &str,
         project_id: &str,
         instance_name: &str,
         is_read_only: bool,
-        channel_size: usize,
         timeout: Option<Duration>,
     ) -> Result<Self> {
         info!("Connecting to bigtable emulator at {}", emulator_endpoint);
-        let endpoints: Vec<Endpoint> = vec![0; channel_size.max(1)]
-            .iter()
-            .map(move |_| {
-                Channel::from_shared(format!("http://{}", emulator_endpoint))
-                    .expect("Invalid connection emulator uri")
-                    .http2_keep_alive_interval(Duration::from_secs(60))
-                    .keep_alive_while_idle(true)
-            })
-            .map(|ep| {
-                if let Some(timeout) = timeout {
-                    ep.timeout(timeout)
-                } else {
-                    ep
-                }
-            })
-            .collect();
+
+        // Parse emulator_endpoint. Officially, it's only host:port,
+        // but unix:///path/to/unix.sock also works in the Go SDK at least.
+        // Having the emulator listen on unix domain sockets without ip2unix is
+        // covered in https://github.com/googleapis/google-cloud-go/pull/9665.
+
+        let endpoint = if emulator_endpoint.starts_with("unix://") {
+            // the URL doesn't matter, we use a custom connector later.
+            Endpoint::from_static("http://[::]:50051")
+        } else {
+            Channel::from_shared(format!("http://{}", emulator_endpoint))
+                .expect("invalid connection emulator uri")
+        };
+
+        let endpoint = endpoint
+            .http2_keep_alive_interval(Duration::from_secs(60))
+            .keep_alive_while_idle(true);
+        let endpoint = if let Some(timeout) = timeout {
+            endpoint.timeout(timeout)
+        } else {
+            endpoint
+        };
+
+        let channel = if let Some(path) = emulator_endpoint.strip_prefix("unix://") {
+            let path: String = path.to_string();
+            let connector = tower::service_fn({
+                move |_: tonic::transport::Uri| UnixStream::connect(path.clone())
+            });
+
+            endpoint.connect_with_connector_lazy(connector)
+        } else {
+            endpoint.connect_lazy()
```

Comment:
I think if you change `fn create_client()` to only take in a channel (without the endpoint vector), then in this else block you can make the usual http endpoint channel, so no need to create the `if emulator_endpoint.starts_with("unix://") {}` block above.

What do you think?

Reply:
This was done.